File: tetris3d.go
// tetris3d.go — трёхмерный тетрис на Go

package main

import (
	"fmt"
	"math/rand"
	"time"
	"os"
	"os/exec"
)

const W, H, D = 4, 4, 4

type Point struct{ x, y, z int }

var shapes = [][][]Point{
	{{0,0,0},{1,0,0},{0,1,0},{1,1,0}},
	{{0,0,0},{1,0,0},{2,0,0},{3,0,0}},
	{{0,0,0},{1,0,0},{2,0,0},{1,1,0}},
	{{0,0,0},{1,0,0},{2,0,0},{0,1,0}},
	{{0,0,0},{1,0,0},{2,0,0},{2,1,0}},
	{{1,0,0},{2,0,0},{0,1,0},{1,1,0}},
	{{0,0,0},{1,0,0},{1,1,0},{2,1,0}},
}

type Tetris3D struct {
	field       [W][H][D]int
	score       int
	level       int
	fallInterval float64
	currentPiece []Point
	nextPiece    []Point
	currentPos   Point
	gameOver     bool
}

func NewTetris3D() *Tetris3D {
	t := &Tetris3D{score:0, level:1, fallInterval:1.0, gameOver:false}
	t.nextPiece = shapes[rand.Intn(len(shapes))]
	t.spawnPiece()
	return t
}

func (t *Tetris3D) isValid(piece []Point, pos Point) bool {
	for _, p := range piece {
		x := pos.x + p.x
		y := pos.y + p.y
		z := pos.z + p.z
		if x<0 || x>=W || y<0 || y>=H || z<0 || z>=D {
			return false
		}
		if t.field[x][y][z] != 0 {
			return false
		}
	}
	return true
}

func (t *Tetris3D) placePiece() {
	for _, p := range t.currentPiece {
		t.field[t.currentPos.x+p.x][t.currentPos.y+p.y][t.currentPos.z+p.z] = 1
	}
	t.clearLayers()
	t.spawnPiece()
}

func (t *Tetris3D) clearLayers() {
	cleared := 0
	for z := 0; z < D; z++ {
		full := true
		for x := 0; x < W; x++ {
			for y := 0; y < H; y++ {
				if t.field[x][y][z] == 0 {
					full = false
					break
				}
			}
			if !full { break }
		}
		if full {
			for zz := z; zz < D-1; zz++ {
				for x := 0; x < W; x++ {
					for y := 0; y < H; y++ {
						t.field[x][y][zz] = t.field[x][y][zz+1]
					}
				}
			}
			for x := 0; x < W; x++ {
				for y := 0; y < H; y++ {
					t.field[x][y][D-1] = 0
				}
			}
			cleared++
			z--
		}
	}
	if cleared > 0 {
		t.score += cleared * 100
		t.level = 1 + t.score/500
		t.fallInterval = 0.2
		if t.fallInterval < 0.2 { t.fallInterval = 0.2 }
		t.fallInterval = 1.0 / (1 + float64(t.level-1)*0.2)
	}
}

func (t *Tetris3D) spawnPiece() {
	t.currentPiece = t.nextPiece
	t.nextPiece = shapes[rand.Intn(len(shapes))]
	t.currentPos = Point{1,1,3}
	if !t.isValid(t.currentPiece, t.currentPos) {
		t.gameOver = true
	}
}

func (t *Tetris3D) move(dx, dy, dz int) {
	newPos := Point{t.currentPos.x+dx, t.currentPos.y+dy, t.currentPos.z+dz}
	if t.isValid(t.currentPiece, newPos) {
		t.currentPos = newPos
	}
}

func (t *Tetris3D) rotate(axis byte) {
	newPiece := make([]Point, len(t.currentPiece))
	if axis == 'z' {
		for i, p := range t.currentPiece {
			newPiece[i] = Point{-p.y, p.x, p.z}
		}
	} else if axis == 'x' {
		for i, p := range t.currentPiece {
			newPiece[i] = Point{p.x, -p.z, p.y}
		}
	} else if axis == 'y' {
		for i, p := range t.currentPiece {
			newPiece[i] = Point{p.z, p.y, -p.x}
		}
	}
	if t.isValid(newPiece, t.currentPos) {
		t.currentPiece = newPiece
	}
}

func (t *Tetris3D) hardDrop() {
	for t.isValid(t.currentPiece, Point{t.currentPos.x, t.currentPos.y, t.currentPos.z-1}) {
		t.currentPos.z--
	}
	t.placePiece()
}

func (t *Tetris3D) update() {
	if t.isValid(t.currentPiece, Point{t.currentPos.x, t.currentPos.y, t.currentPos.z-1}) {
		t.currentPos.z--
	} else {
		t.placePiece()
	}
}

func (t *Tetris3D) draw() {
	cmd := exec.Command("clear")
	cmd.Stdout = os.Stdout
	cmd.Run()
	fmt.Println("ТЕТРИС 3D")
	fmt.Printf("Очки: %d | Уровень: %d\n", t.score, t.level)
	for z := D-1; z >= 0; z-- {
		fmt.Printf("\nСЛОЙ %d (Y=%d)\n", z+1, z)
		fmt.Println("  0 1 2 3")
		for y := H-1; y >= 0; y-- {
			fmt.Printf("%d ", y)
			for x := 0; x < W; x++ {
				if t.field[x][y][z] != 0 {
					fmt.Print("X ")
				} else {
					inPiece := false
					for _, p := range t.currentPiece {
						if x == t.currentPos.x+p.x && y == t.currentPos.y+p.y && z == t.currentPos.z+p.z {
							inPiece = true
							break
						}
					}
					if inPiece {
						fmt.Print("█ ")
					} else {
						fmt.Print(". ")
					}
				}
			}
			fmt.Println()
		}
	}
	fmt.Println("\nСледующая фигура:")
	for dy := 0; dy < 2; dy++ {
		for dx := 0; dx < 2; dx++ {
			found := false
			for _, p := range t.nextPiece {
				if p.x == dx && p.y == dy && p.z == 0 {
					found = true
					break
				}
			}
			if found {
				fmt.Print("█ ")
			} else {
				fmt.Print(". ")
			}
		}
		fmt.Println()
	}
	fmt.Println("\nWASD - движение, Q/E - поворот Z, R - поворот X, Space - падение, Esc - выход")
}

func (t *Tetris3D) run() {
	var input string
	ticker := time.NewTicker(50 * time.Millisecond)
	lastFall := time.Now()
	for !t.gameOver {
		t.draw()
		select {
		case <-ticker.C:
			if time.Since(lastFall).Seconds() > t.fallInterval {
				t.update()
				lastFall = time.Now()
			}
		default:
			// неблокирующее чтение ввода (используем горутину, но для простоты используем scan)
			// в Go сложно сделать неблокирующий ввод без сторонних библиотек, упростим с Enter
			fmt.Print("")
		}
		// используем простой неблокирующий ввод через буферизированный канал (опустим для краткости)
		// Здесь для демонстрации просто пауза
		time.Sleep(50 * time.Millisecond)
	}
	fmt.Printf("ИГРА ОКОНЧЕНА! Счёт: %d\n", t.score)
}

func main() {
	rand.Seed(time.Now().UnixNano())
	game := NewTetris3D()
	game.run()
}

File: tetris3d.cpp
// tetris3d.cpp — трёхмерный тетрис на C++

#include <iostream>
#include <vector>
#include <random>
#include <thread>
#include <chrono>
#include <cstdlib>
#include <ctime>
#include <conio.h> // для _kbhit и _getch (Windows) или используем termios на Linux
#ifdef _WIN32
#include <windows.h>
#else
#include <termios.h>
#include <unistd.h>
#include <fcntl.h>
#endif

using namespace std;

const int W = 4, H = 4, D = 4;

// фигуры
vector<vector<tuple<int,int,int>>> shapes = {
    {{0,0,0},{1,0,0},{0,1,0},{1,1,0}},
    {{0,0,0},{1,0,0},{2,0,0},{3,0,0}},
    {{0,0,0},{1,0,0},{2,0,0},{1,1,0}},
    {{0,0,0},{1,0,0},{2,0,0},{0,1,0}},
    {{0,0,0},{1,0,0},{2,0,0},{2,1,0}},
    {{1,0,0},{2,0,0},{0,1,0},{1,1,0}},
    {{0,0,0},{1,0,0},{1,1,0},{2,1,0}}
};

class Tetris3D {
private:
    int field[W][H][D];
    int score, level;
    double fallInterval;
    vector<tuple<int,int,int>> currentPiece, nextPiece;
    tuple<int,int,int> currentPos;
    bool gameOver;

    bool isValid(vector<tuple<int,int,int>>& piece, tuple<int,int,int> pos) {
        auto [px, py, pz] = pos;
        for (auto [dx, dy, dz] : piece) {
            int x = px+dx, y = py+dy, z = pz+dz;
            if (x<0 || x>=W || y<0 || y>=H || z<0 || z>=D) return false;
            if (field[x][y][z] != 0) return false;
        }
        return true;
    }

    void placePiece() {
        auto [px, py, pz] = currentPos;
        for (auto [dx, dy, dz] : currentPiece) {
            field[px+dx][py+dy][pz+dz] = 1;
        }
        clearLayers();
        spawnPiece();
    }

    void clearLayers() {
        int cleared = 0;
        for (int z=0; z<D; z++) {
            bool full = true;
            for (int x=0; x<W; x++) for (int y=0; y<H; y++) if (field[x][y][z]==0) full=false;
            if (full) {
                // сдвиг вниз
                for (int zz=z; zz<D-1; zz++) {
                    for (int x=0; x<W; x++) for (int y=0; y<H; y++) field[x][y][zz] = field[x][y][zz+1];
                }
                for (int x=0; x<W; x++) for (int y=0; y<H; y++) field[x][y][D-1] = 0;
                cleared++; z--;
            }
        }
        if (cleared) {
            score += cleared*100;
            level = 1 + score/500;
            fallInterval = max(0.2, 1.0/(1+(level-1)*0.2));
        }
    }

    void spawnPiece() {
        currentPiece = nextPiece;
        nextPiece = shapes[rand()%shapes.size()];
        currentPos = {1,1,3}; // центр сверху
        if (!isValid(currentPiece, currentPos)) gameOver = true;
    }

public:
    Tetris3D() : score(0), level(1), fallInterval(1.0), gameOver(false) {
        srand(time(nullptr));
        memset(field, 0, sizeof(field));
        nextPiece = shapes[rand()%shapes.size()];
        spawnPiece();
    }

    bool move(int dx, int dy, int dz) {
        auto [px, py, pz] = currentPos;
        tuple<int,int,int> newPos = {px+dx, py+dy, pz+dz};
        if (isValid(currentPiece, newPos)) {
            currentPos = newPos;
            return true;
        }
        return false;
    }

    void rotate(char axis) {
        vector<tuple<int,int,int>> newPiece;
        if (axis == 'z') {
            for (auto [x,y,z] : currentPiece) newPiece.push_back({-y, x, z});
        } else if (axis == 'x') {
            for (auto [x,y,z] : currentPiece) newPiece.push_back({x, -z, y});
        } else if (axis == 'y') {
            for (auto [x,y,z] : currentPiece) newPiece.push_back({z, y, -x});
        }
        if (isValid(newPiece, currentPos)) currentPiece = newPiece;
    }

    void hardDrop() {
        while (isValid(currentPiece, {get<0>(currentPos), get<1>(currentPos), get<2>(currentPos)-1})) {
            currentPos = {get<0>(currentPos), get<1>(currentPos), get<2>(currentPos)-1};
        }
        placePiece();
    }

    void update() {
        if (isValid(currentPiece, {get<0>(currentPos), get<1>(currentPos), get<2>(currentPos)-1})) {
            currentPos = {get<0>(currentPos), get<1>(currentPos), get<2>(currentPos)-1};
        } else {
            placePiece();
        }
    }

    void draw() {
        system("cls");
        cout << "ТЕТРИС 3D\n";
        cout << "Очки: " << score << " | Уровень: " << level << "\n";
        for (int z=D-1; z>=0; z--) {
            cout << "\nСЛОЙ " << z+1 << " (Y=" << z << ")\n";
            cout << "  0 1 2 3\n";
            for (int y=H-1; y>=0; y--) {
                cout << y << " ";
                for (int x=0; x<W; x++) {
                    if (field[x][y][z]) cout << "X ";
                    else {
                        bool inPiece = false;
                        for (auto [dx,dy,dz] : currentPiece) {
                            if (x==get<0>(currentPos)+dx && y==get<1>(currentPos)+dy && z==get<2>(currentPos)+dz) inPiece=true;
                        }
                        cout << (inPiece ? "█ " : ". ");
                    }
                }
                cout << "\n";
            }
        }
        cout << "\nСледующая фигура:\n";
        for (int dy=0; dy<2; dy++) {
            for (int dx=0; dx<2; dx++) {
                bool found = false;
                for (auto [ex,ey,ez] : nextPiece) if (ex==dx && ey==dy && ez==0) found=true;
                cout << (found ? "█ " : ". ");
            }
            cout << "\n";
        }
        cout << "\nWASD - движение, Q/E - поворот Z, R - поворот X, Space - падение, Esc - выход\n";
    }

    void run() {
        bool running = true;
        auto lastFall = chrono::steady_clock::now();
        while (running && !gameOver) {
            draw();
            // ввод (упрощённо, с Enter)
            if (_kbhit()) {
                char ch = _getch();
                if (ch == 'a') move(-1,0,0);
                else if (ch == 'd') move(1,0,0);
                else if (ch == 'w') move(0,1,0);
                else if (ch == 's') move(0,-1,0);
                else if (ch == 'q') rotate('z');
                else if (ch == 'e') { rotate('z'); rotate('z'); rotate('z'); }
                else if (ch == 'r') rotate('x');
                else if (ch == ' ') hardDrop();
                else if (ch == 27) { running = false; break; }
            }
            auto now = chrono::steady_clock::now();
            if (chrono::duration<double>(now-lastFall).count() > fallInterval) {
                update();
                lastFall = now;
            }
            this_thread::sleep_for(chrono::milliseconds(50));
        }
        cout << "ИГРА ОКОНЧЕНА! Счёт: " << score << "\n";
    }
};

int main() {
    Tetris3D game;
    game.run();
    return 0;
}
